backend/crud.py

```diff
@@ -0,0 +1,211 @@
+from typing import Dict
+from sqlalchemy.orm import Session
+from sqlalchemy import select, update
+
+from .models import (
+    AttrType,
+    Attribute,
+    AttributeDefinition,
+    BoundFK,
+    Entity,
+    Schema,
+    Value,
+    val_map,
+    type_map
+)
+
+from .schemas import (
+    SchemaCreateSchema,
+    SchemaUpdateSchema,
+    AttributeCreateSchema
+)
+
+
+def create_attribute(db: Session, data: AttributeCreateSchema) -> Attribute:
+    exists = db.execute(select(Attribute).where(Attribute.name == data.name)).scalar()
+    if exists:
+        # TODO custom exceptions
+        raise Exception(f'Attribute with name `{data.name}` already exists')
+    a = Attribute(name=data.name, type=data.type)
+    db.add(a)
+    db.commit()
+    return a
+
+
+def create_schema(db: Session, data: SchemaCreateSchema) -> Schema:
+    # TODO require lock here and in update?
+    exists = db.execute(select(Schema).where( (Schema.name == data.name) | (Schema.slug == data.slug) )).scalar()
+    if exists:
+        raise Exception(f'Schema with name `{data.name}` or slug `{data.slug}` already exists')
+    
+    sch = Schema(name=data.name, slug=data.slug)
+    db.add(sch)
+    db.flush()
+    for attr in data.attributes:
+        a: Attribute = db.execute(select(Attribute).where(Attribute.id == attr.attr_id)).scalar()
+        if a is None:
+            raise Exception(f'Attribute with id {attr.attr_id} does not exist')
+        ad = AttributeDefinition(
+            attribute_id=attr.attr_id, 
+            schema_id=sch.id,
+            required=attr.required, 
+            list=attr.list, 
+            unique=attr.unique,
+            key=attr.key,
+            description=attr.description
+        )
+        db.add(ad)
+        db.flush()
+        if a.type == AttrType.fk:
+            if attr.bind_to_schema is None:
+                raise Exception('You need to bind foreign key to some schema')
+            if attr.bind_to_schema == -1:
+                s = sch
+            else:
+                s = db.execute(select(Schema).where(Schema.id == attr.bind_to_schema)).scalar()
+            if s is None:
+                raise Exception(f'Cannot bind to nonexistent schema with id {attr.bind_to_schema}')
+            bfk = BoundFK(attr_def_id=ad.id, schema_id=s.id)
+            db.add(bfk)
+    db.commit()
+    return sch
+
+
+def delete_schema(db: Session, schema_id: int) -> Schema:
+    schema = db.execute(select(Schema).where(Schema.id == schema_id)).scalar()
+    if schema is None:
+        raise Exception('Should we raise exception here?')
```

Comment:
I would assume that a deletion of a schema will only be triggered by an API request. In this case, having a status code 404 seems to be a worthy goal. To achieve this, an exception seems like a good choice :+1:

Reply:
You might also want to check, whether the schema is already deleted. In that case a 404 response would also be warranted.